app/objects/c_ability.py

```diff
@@ -140,25 +89,23 @@ async def which_plugin(self):
                 return plugin
         return None
 
-    def replace_cleanup(self, encoded_cmd, payload):
-        decoded_cmd = b64decode(encoded_cmd).decode('utf-8', errors='ignore').replace('\n', '')
-        decoded_cmd = decoded_cmd.replace(self.RESERVED['payload'], payload)
-        return decoded_cmd
+    def get_executor(self, platform, name):
+        return next(iter([e for e in self.executors if e.platform == platform and e.name == name]), None)
+
+    def get_executors(self, platform, names):
+        executors = []
+        for name in names:
+            matched_executor = self.get_executor(platform, name)
+            if matched_executor and matched_executor not in executors:
+                executors.append(matched_executor)
+        return executors
+
+    def add_executor(self, executor):
+        existing_executor = self.get_executor(executor.platform, executor.name)
+        if existing_executor:
+            self.executors.remove(existing_executor)
+        self.executors.append(executor)
```

Comment:
Why do we remove just to re-add it?

Reply:
This removes any executor that matches the incoming executor's platform and name. Abilities should have unique executors for each platform/executor combination.

Ability files _could_ have something like the following, but we only want to store one executor (the last one):

```
- id: uuid-goes-here
   ...
   platforms:
     windows:
       psh:
         command: whoami
    windows:
       psh:
         command: env
```

Open to suggestions.

Reply:
I missed some indentation in the YAML. But yeah you're right that either the parser would blow up or the last one in would win. 

I think it would be appropriate to keep this in the `add_executor(executor)` method, to ensure uniqueness if new executors are added (unless someone interacts directly with `executors`).

I'm not sure how a set would work here. Two `Executor` objects are created with the same platform/name, they would still be unique objects. Is there a way to handle this?

Reply:
We talked offline and it might be useful to use a dictionary to organize the executors internally and maintain uniqueness. Adding/removing executors would be done through defined interfaces (like `add_executor()`) and direct access to the internal store could be disallowed.

Reply:
If going with the internal dict, for backwards compatibility in terms of reading (and maybe for convenience) you could have a `@property` like the following:

```python
@property
def executors(self):
    for executor in self._executor_map.values():
        yield executor
```